torch/csrc/autograd/python_function.cpp

```diff
@@ -366,100 +349,79 @@ static void _wrap_outputs(THPFunction *self, t2var_type &t2var,
-    }
-    if (is_modified) {
-      if (prev.is_leaf() && prev.requires_grad()) {
+      var.detach_();
+    } else if (is_modified) {
+      if (var.is_leaf() && var.requires_grad()) {
         throw std::runtime_error("a leaf Variable that requires grad has been used in an in-place operation.");
       }
       // If the input was modified, transplant the grad_fn in the graph:
       // grad_fn <- variable <- self  ==>  grad_fn <- self <- variable
-      prev.get()->grad.reset();
-      prev.get()->hooks.clear();
-      if (auto grad_acc_fn = prev.get()->grad_accumulator.lock()) {
+      var.get()->grad.reset();
+      var.get()->hooks.clear();
+      if (auto grad_acc_fn = var.get()->grad_accumulator.lock()) {
         auto grad_acc = dynamic_cast<AccumulateGrad*>(grad_acc_fn.get());
         grad_acc->variable.reset();
       }
       if (cdata) {
-        prev.rebase_history(output_nr, cdata);
+        var.rebase_history(output_nr, cdata);
       }
-      return prev;
+    } else if (is_input) {
+      // An input has been returned, but it wasn't modified. Return it as a view
+      // so that we can attach a new grad_fn to the Variable.
+      var = var.slice();
+      var.get()->output_nr = output_nr;
+      var.get()->_grad_fn = cdata;
+    } else if (cdata) {
+      var.get()->output_nr = output_nr;
+      var.get()->_grad_fn = cdata;
     }
-    // An input has been returned, but it wasn't modified. Return it as a view
-    // so that we can attach a new grad_fn to the Variable.
-    return make_variable_view(std::move(prev), std::move(data), output_nr, cdata);
   };
 
-  t2var_type output2var;
   for (int i = 0; i < num_outputs; i++) {
-    PyObject *output = PyTuple_GET_ITEM(raw_output, i);
+    PyObject* obj = PyTuple_GET_ITEM(raw_output, i);
 
-    THPVariable* output_var;
-    auto it = output2var.find(output);
-    if (it != output2var.end()) {
-      output_var = it->second;
-      Py_INCREF(output_var);
-    } else {
-      // Wrap the output in a Variable
-      bool is_modified = dirty_inputs.count(output) > 0;
-      bool is_non_differentiable = non_differentiable.count(output) > 0;
-      Variable var = wrap_output(
-          torch::createTensor(output),
-          get_shared_base(output),
-          i,
-          is_modified,
-          is_non_differentiable);
-
-      output_var = (THPVariable*)THPVariable_Wrap(var);
-      if (!output_var) throw python_error();
-
-      // We already have the data tensor wrapped as a PyObject*
-      Py_INCREF(output);
-      Py_CLEAR(output_var->data);
-      output_var->data = output;
-
-      output2var[output] = output_var;
-    }
+    bool is_input = inputs.count(obj) > 0;
+    bool is_modified = std::find(dirty_inputs.begin(), dirty_inputs.end(), obj) != dirty_inputs.end();
+    bool is_non_differentiable = non_differentiable.count(obj) > 0;
+
+    auto var = as_variable(obj, i);
+    set_history(var, i, is_input, is_modified, is_non_differentiable);
```

Comment:
Can you tell me why is it the case that we no longer need to track duplication? We will just destroy the previous history assignment in here. In general dealing with repeated outputs is hard, but I think the first one should be a proper Variable, and all other ones should be implicit views, so they can get gradient too. Right now one of `grad_outputs` for a repeated output is guaranteed to be None

Reply:
I think we should prefer returning the same object when possible. I've added explicit tests for this, although other tests triggered this code path too.

When var is a view, this raises an exception because we don't support in-place modifications to views from functions that return multiple values.

Otherwise, the last call to set_history wins. All that will change is the `output_nr` which doesn't really matter.

Reply:
Hmm idk, forbiding that would be safer. Do we depend on it anywhere outside of our test cases? Also, can you add a comment saying that it's a deliberate decision?